[AArch64] Allow LDR merge with same destination register by renaming

nocchijiang opened the pull request:
The patch is based on a reverted patch: https://reviews.llvm.org/D103597. It was trying to rename registers before alias check, which is not safe and causes miscompiles. This patch does 2 things:

1. Do the renaming with necessary checks passed, including alias check.
2. Rename the register for the instructions between the pairs and combine the second load into the first. By doing so we can just check the renamability between the pairs and avoid scanning unknown amount of instructions before/after the pairs.

Necessary refactoring has been made in order to reuse as much code possible with STR renaming.

nocchijiang committed:
[AArch64] Allow LDR merge with same destination register by renaming

The patch is based on a reverted patch: https://reviews.llvm.org/D103597.
It was trying to rename registers before alias check, which is not safe
and causes miscompiles. This patch does 2 things:

1. Do the renaming with necessary checks passed, including alias check.
2. Rename the register for the instructions between the pairs and
   combine the second load into the first. By doing so we can just check
   the renamability between the pairs and avoid scanning unknown amount
   of instructions before/after the pairs.

Necessary refactoring has been made in order to reuse as much code as
possible with STR renaming.

nocchijiang commented:
@mstorsjo @sjoerdmeijer @fhahn @MeeraN7 

Since I don't have commit access, please help add reviewers for me and take a look at the PR when you are free. Thanks!

davemgreen commented:
Hi. This looks like an interesting patch. Have you tried running a bootstrap or the llvm-test-suite, to make sure it behaves OK? Thanks

nocchijiang commented:
> Have you tried running a bootstrap or the llvm-test-suite, to make sure it behaves OK?

I tested the patch by running llvm-test-suite before rebasing onto the tip of `main`. I just found that the test was broken by the [patch](https://reviews.llvm.org/D88663) landed 2 weeks ago, which is fixed by the commit I just pushed. @c-rhodes please also have a look at this to make sure I'm not breaking anything.

I'm not familiar with bootstrapping stuff. I'm working on an Apple Silicon Mac so I followed [this guide](https://llvm.org/docs/AdvancedBuilds.html#apple-clang-builds-a-more-complex-bootstrap) to build Apple clang but no luck:

```
ninja: error: unknown target 'stage2-distribution'
```

nocchijiang commented:
Rebased onto latest `main` to resolve conflicts.

nocchijiang commented on llvm/lib/Target/AArch64/AArch64LoadStoreOptimizer.cpp:
```diff
@@ -815,12 +815,11 @@ AArch64LoadStoreOpt::mergeNarrowZeroStores(MachineBasicBlock::iterator I,
-                              const TargetRegisterInfo *TRI, unsigned Limit,
-                              std::function<bool(MachineInstr &, bool)> &Fn) {
+static bool forAllMIsUntil(MachineInstr &MI,
+                           std::function<bool(MachineInstr &MI, bool)> Until,
```
Sorry for the confusion made here. Initially I wanted to let `Until` decide whether to stop the reverse iteration (because the exit point is fixed for merging `ldr` - right at the first load). Later I realized that the first load must be the def thus the entire functions can be reused by just iterating from different MIs for store and load, leaving `Until` unused. I have corrected this in the latest commit.

davemgreen reviewed (approved):
Thanks. Whenever we try these kind of things there seems to often be a high chance of problems occurring. Please make sure you have tested thoroughly, but from what I can tell this LGTM.

nocchijiang commented:
Rebased, squashed all commits and addressed comments.

@davemgreen Since I don't have commit access yet, could you please merge the PR for me? Before pushing the update, I have tested the patch by compiling Chromium using the bootstrapped Clang, passing all the unit tests from Chromium on an Apple Silicon Mac.

davemgreen commented:
Thanks. Sounds good.
